lib/components/viewers/stop-schedule-table.js

```diff
@@ -0,0 +1,152 @@
+import styled from 'styled-components'
+
+import { isBlank } from '../../util/ui'
+import {
+  excludeLastStop,
+  getRouteIdForPattern,
+  getSecondsUntilDeparture,
+  getStopTimesByPattern,
+  routeIsValid,
+  stopTimeComparator
+} from '../../util/viewer'
+import StopTimeCell from './stop-time-cell'
+
+// Styles for the schedule table and its contents.
+const StyledTable = styled.table`
+  border-spacing: collapse;
+  height: 100%;
+  width: 100%;
+  th {
+    background-color: #fff;
+    box-shadow: 0 1px 0px 0px #ccc;
+    font-size: 75%;
+    position: sticky;
+    top: 0px;
+  }
+  tr > * {
+    border-bottom: 1px solid #ccc;
+    padding: 2px 0 2px 10px;
+    vertical-align: top;
+  }
+  td:first-child, th:first-child {
+    padding-left: 0;
+  }
+`
+
+const RouteCell = styled.td`
+  font-weight: bold;
+`
+const TimeCell = styled.td`
+  font-weight: bold;
+  text-align: right;
+  white-space: nowrap;
+`
+
+class StopScheduleTable extends Component {
+  firstDepartureRef = createRef()
+
+  /*
+   * Scroll to the first stop time that is departing from now.
+   */
+  _scrollToFirstDeparture = () => {
+    const { current } = this.firstDepartureRef
+    if (current) {
+      current.scrollIntoView()
+    }
+  }
+
+  componentDidMount () {
+    this._scrollToFirstDeparture()
+  }
+
+  componentDidUpdate () {
+    // Should only happen if user changes date and a new stopData is passed.
+    this._scrollToFirstDeparture()
+  }
+
+  render () {
+    const { homeTimezone, stopData, timeFormat } = this.props
+    const stopTimesByPattern = getStopTimesByPattern(stopData)
+
+    // Merge stop times, so that we can sort them across all route patterns.
+    // (stopData is assumed valid per StopViewer render condition.)
+    let mergedStopTimes = []
+    Object.values(stopTimesByPattern).forEach(({ pattern, route, times }) => {
+      // Only add pattern if route info is returned by OTP.
+      if (routeIsValid(route, getRouteIdForPattern(pattern))) {
+        const filteredTimes = times
+          .filter(excludeLastStop)
+          .map(stopTime => {
+            // Add the route attribute and headsign to each stop time for rendering route info.
+            const headsign = isBlank(stopTime.headsign) ? pattern.headsign : stopTime.headsign
+            return {
+              ...stopTime,
+              route,
+              headsign
+            }
+          })
+        mergedStopTimes = mergedStopTimes.concat(filteredTimes)
+      }
+    })
+
+    mergedStopTimes = mergedStopTimes.sort(stopTimeComparator)
+
+    // Find the next stop time that is departing. We will scroll to that stop time entry.
+    let firstDepartureFromNow
+    if (mergedStopTimes.length) {
+      // Search starting from the last stop time (largest seconds until departure) for this pattern.
+      const lastStopTime = mergedStopTimes[mergedStopTimes.length - 1]
+
+      firstDepartureFromNow = mergedStopTimes.reduce((firstStopTime, stopTime) => {
```

Comment:
I'm not a big fan of reduce functions -- I don't find them very user friendly. Could you look into using a different approach?

Reply:
I extracted the method and made it use a plain `forEach` in https://github.com/opentripplanner/otp-react-redux/pull/292/commits/edd5a4bf88907df9d012180bedebaaddbc01d6f8.